### ISSUE TITLE
slow apt search with lz4 indexes

### ISSUE THREAD

**Issue by jfikar** [Debian/apt#115]
I've recently found very slow performance of `apt search test` on aarch64 with Ubuntu 20.04. The indexes are compressed by lz4 as specified in /etc/apt/apt.conf.d/02-armbian-compress-indexes

```
Acquire::GzipIndexes "true";
Acquire::CompressionTypes::Order:: "gz";
```
It seems that only the first line matters. When it is false, the indexes are not compressed. I'm not sure, how the second line influences the compression. It seems it does not matter.

By default, i.e. lz4 compressed indexes, the `time apt search test` takes 15 minutes. One core is fully loaded. That's too much.

When I decompress the indexes, or change GzipIndexes to "false", it takes 8 seconds.

When I recompress the lz4 files to lzo by lzop, it takes only 0.4 second.

I suspect something is wrong handling the lz4 files, not sure what.

When I recompress the lz4 files by hand to gz, the `time apt search test` takes 81 minutes. Apparently gz is also slow.

I've found a [recent discussion](https://forum.armbian.com/topic/14064-my-apt-search-has-become-super-slow-recently/), which  directed me to the /etc/apt/apt.conf.d/02-armbian-compress-indexes file, but it does not explain the difference in fast lzo and slow lz4 and gz indexes.

Also, GzipIndexes seems a bit misleading.

**Comment by DonKult** [comment]
The second config option influences in which compression format a file is downloaded (if available). APT usually defaults to downloading as xz (that is based on the inbuilt cost-rating each supported compression has). update will uncompress the files (e.g. to check hashes) and store them eventually in compressed form (if GzipIndexes is enabled) nowadays usually in lz4 (which has the lowest inbuilt cost-rating – except uncompressed of course). The misnomer is historic baggage as that used to be only supported with gzip and would not recompress. That said, it is a legacy option as this is better configured per target and some (like Contents) do, while for others (like Packages and Translation-*) there would actually need to be still some work put in to make it work optimally if that option is enabled.

That your lzo test is so fast is a red herring: APT does not support lzo compression, so it is of course very fast in not reading the file… (I assuming you haven't manually added support for it via config, which is out-of-scope for this comment. The manpage has some details if you really really want to add compressors at runtime).

So, long story short: We aren't the ones who have set these options for you. If you think they shouldn't be set in your environment and usecase please complain to those who set it, not to us.

Honestly: That compressed files are slow is kinda expected, especially for a seek-heavy usecase like search. You gain less filesize in exchange. So far we as upstream consider the trade-off not worth it, for these files as global default, but downstreams with more specific usecases can have other opinions. It is a trade-off after all.

**Comment by jfikar** [comment]
I see, thank you for explanation. Removing the compression was the solution in armbian forum.

The strange thing is that with lzo indexes the apt search still seems to return correct answers. So my apt 2.0.2ubuntu0.1 (arm64) can also read lzo files and can handle them very fast.

So what is the difference in handling lzo and lz4 files? Maybe lzo is first decompressed by an external program and then searched? While lz4 is build-in, but slow? Can it be that removing lz4 build-in support will speed up handling lz4 indexes?

I understand that seek in compressed files is expected to be slow, but the difference in speed between lzo and lz4 is huge. The factor is almost 2000x. Lzo is even a bit faster than the uncompressed indexes.

**Comment by DonKult** [comment]
My guess is that the "correct" result you see is based on files which are not compressed. Like the packages you have installed (/var/lib/dpkg/status).

APT can certainly be configured to use external binaries to (de)compress files as hinted already (my comment about the manpage having the details), but I kinda doubt you/armbian having done that for lzo. See `apt-config dump APT::Compressor` for which are supported & how they are configured. Note that this will show how apt would call e.g. gzip (if you have it installed) even if it will use the library (if it was built against it, which it is by default). On the other hand, if you don't have the binary, but we have library linked, it will show binary "false". Probably e.g. the case for lz4 for you. Not having APT::Compressor::lzo means that APT is not using lzo files which explains why its so fast…

(It is possible that a compressed file is faster than an uncompressed if your disk is slow and the CPU fast enough but I kinda doubt it is the case here as Packages files are too small. On Contents files you could probably see that…)

**Comment by jfikar** [comment]
You are completely right. In fact, the fast search with lzo indexes returns much less results than the search with no compression or lz4. So it does not search the indexes. Also, `apt-config dump APT::Compressor` does not show lzo. There is zstd, so I've tested that, but the speed 42 minutes is somewhere between lz4 15 minutes and gzip 81 minutes (as expected).

Now it is clear, I'm switching the compression off.
Thank you and sorry for the noise
